File: haskell-server/src/EagerWords/Server/Domain/Scorer.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- TODO. Create tests for the scorer module.
-- Pay particular attention to representation of empty cells in the implementation.
-- Then connect this module to the rest of the application and use it to compute scores.
module EagerWords.Server.Domain.Scorer (
    Scorer
  , Scorer(scorePlay, scoreWord)
  , mkScorer
  ) where

import qualified Data.List as List
import Bolour.Plane.Domain.Point (Point, Point(Point))
import qualified Bolour.Plane.Domain.Point as Point
import EagerWords.Common.Domain.PlayPiece (PlayPiece, PlayPiece(PlayPiece), MoveInfo)
import qualified EagerWords.Common.Domain.PlayPiece as PlayPiece

data Scorer = Scorer {
    scorePlay :: [PlayPiece] -> Int
  , scoreWord :: [PlayPiece] -> Int
}

mkScorer :: [[Int]] -> Scorer
mkScorer pointValues =
  Scorer
       (doScoreWord pointValues)
       (doScoreWord pointValues)

doScoreWord :: [[Int]] -> [PlayPiece] -> Int
doScoreWord pointValues playPieces =
  let movedPoints = PlayPiece.point <$> List.filter PlayPiece.moved playPieces
      value Point {row, col} = (pointValues !! row) !! col
      accumulate total point = total + value point
  in List.foldl' accumulate 0 movedPoints

File: scala-server/app/com/bolour/eagerwords/server/domain/Scorer.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */

package com.bolour.eagerwords.server.domain

import com.bolour.eagerwords.common.domain.PlayPiece
import com.bolour.plane.domain.Point
import org.slf4j.LoggerFactory

/**
  * Scores plays.
  *
  * @param pointValues 2D list of values of the board's points.
  */
class Scorer(pointValues: List[List[Int]]) {

  val logger = LoggerFactory.getLogger(this.getClass)

  // val multiplierGrid: Grid[ScoreMultiplier] = mkMultiplierGrid(dimension)

  /**
    * Score a play.
    *
    * @param playPieces Consecutive list of all play pieces for a play.
    *                   Includes both moved and existing pieces forming the play word.
    * @return The score of the play.
    */
  def scorePlay(playPieces: List[PlayPiece]): Int = scoreWord(playPieces)

  // playPieces map { pp => (pp.piece.value, pp.point, pp.moved)

  /**
    * Score an individual word.
    *
    * @param playPieces Consecutive list of all play pieces for a play.
    *                   Includes both moved and existing pieces forming the play word.
    * @return The score of the word.
    */
  def scoreWord(playPieces: List[PlayPiece]): Int = {
    val movedPoints = playPieces filter { _.moved } map {_.point}
    def value: Point => Int = {case Point(row, col) => pointValues(row)(col)}
    val score = movedPoints.foldLeft(0)((total, point) => total + value(point))
    score
  }
}

object Scorer {
  type Score = Int
  def apply(dimension: Int, trayCapacity: Int, pointValues: List[List[Int]]): Scorer =
    new Scorer(pointValues)
}
